org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JVMConfigurator.java

```diff
@@ -70,19 +78,116 @@ public static boolean configureDefaultVM(Preferences preferences) throws CoreExc
 					}
 					JavaRuntime.setDefaultVMInstall(vm, new NullProgressMonitor());
 					JDTUtils.setCompatibleVMs(vm.getId());
+					changed = true;
+				}
+			}
+		}
+		changed = configureJVMs(preferences, changed);
+		return changed;
+	}
+
+	public static boolean configureJVMs(Preferences preferences, boolean changed) throws CoreException {
+		Set<RuntimeEnvironment> runtimes = preferences.getRuntimes();
+		for (RuntimeEnvironment runtime : runtimes) {
+			if (runtime.isValid()) {
+				File file = runtime.getInstallationFile();
+				if (file != null && file.isDirectory()) {
+					URL javadocURL = runtime.getJavadocURL();
+					IPath sourcePath = runtime.getSourcePath();
+					IVMInstall vm = findVM(file, runtime.getName());
+					IVMInstallType installType = JavaRuntime.getVMInstallType(StandardVMType.ID_STANDARD_VM_TYPE);
+					VMStandin vmStandin;
+					if (vm == null) {
+						long unique = System.currentTimeMillis();
+						while (installType.findVMInstall(String.valueOf(unique)) != null) {
+							unique++;
+						}
+						String vmId = String.valueOf(unique);
+						vmStandin = new VMStandin(installType, vmId);
+					} else {
+						vmStandin = new VMStandin(vm);
+					}
+					IStatus status = installType.validateInstallLocation(file);
+					if (!status.isOK()) {
+						JavaLanguageServerPlugin.log(status);
+						continue;
+					}
+
+					vmStandin.setName(runtime.getName());
+					vmStandin.setInstallLocation(file);
+					if (sourcePath != null || javadocURL != null) {
+						StandardVMType svt = (StandardVMType) installType;
+						LibraryLocation[] libs = svt.getDefaultLibraryLocations(file);
+						if (libs != null) {
+							for (int i = 0; i < libs.length; i++) {
+								LibraryLocation lib = libs[i];
+								if (sourcePath != null) {
+									lib.setSystemLibrarySource(sourcePath);
+								}
+								if (javadocURL != null) {
+									lib = new LibraryLocation(lib.getSystemLibraryPath(), lib.getSystemLibrarySourcePath(), lib.getPackageRootPath(), javadocURL, lib.getIndexLocation(), lib.getExternalAnnotationsPath());
+								}
+								libs[i] = lib;
+							}
+						}
+						vmStandin.setLibraryLocations(libs);
+					}
+					vm = vmStandin.convertToRealVM();
+					if (runtime.isDefault()) {
+						JavaRuntime.setDefaultVMInstall(vm, new NullProgressMonitor());
+					}
+					if (!setDefaultEnvironmentVM(vm, runtime.getName())) {
+						JavaLanguageServerPlugin.logError("Runtime at '" + runtime.getPath() + "' is not compatible with the '" + runtime.getName() + "' environment");
+					}
+					changed = true;
+				} else {
+					JavaLanguageServerPlugin.logInfo("Invalid runtime: " + runtime);
+				}
+			}
+		}
```

Comment:
should log which runtime is the default VM (as multiple ones can have default:true, the last one wins)

Reply:
Done.